tapfreighter/chain_porter.go

```diff
@@ -150,9 +150,20 @@ func (p *ChainPorter) Start() error {
 
 		// Start the main chain porter goroutine.
 		p.Wg.Add(1)
-		go p.mainEventLoop()
+		go func() {
+			defer p.Wg.Done()
+			p.mainEventLoop()
+		}()
 
-		startErr = p.resumePendingParcels()
+		// Resume any pending parcels in a new goroutine so that we
+		// don't delay returning from the Start method. Without this
+		// goroutine the resumePendingParcels method would block on the
+		// buffering constraint of the outboundParcels channel.
+		p.Wg.Add(1)
+		go func() {
+			defer p.Wg.Done()
+			startErr = p.resumePendingParcels()
```

Comment:
I think `resumePendingParcels()` needs to be modified to receive Quit signals?

Like this case:

https://github.com/lightninglabs/taproot-assets/blob/1b5a4ef392dcad4906272b1ef805d7333f5d4080/tapfreighter/chain_porter.go#L330

But here:

https://github.com/lightninglabs/taproot-assets/blob/1b5a4ef392dcad4906272b1ef805d7333f5d4080/tapfreighter/chain_porter.go#L194

Reply:
Why is it even a problem if we delay startup by waiting for pending parcels? Anything that takes a while (e.g. proof transfer) will be done in a goroutine anyway. So I don't see a pressing reason to do things async here.

Also, if we do this in another goroutine, we don't need the buffered channel as already mentioned by @jharveyb. I think it just makes the behavior less deterministic (e.g. with 9 parcels the goroutine finishes almost immediately but with 11 it blocks until complete)...

Reply:
IIUC the start of each 'subsystem' is sequential and blocking on any one `Start()` call would block the rest?

https://github.com/lightninglabs/taproot-assets/blob/a8d8b5a46552c98fe6ef54c796dffa17927b8e33/server.go#L185

Even with proof transfer in a goroutine, if we're resuming more than one parcel I think the transfer process for the first one would block resumption of the second? And these wouldn't happen in parallel.

The concrete case I was thinking of was: "If I have a wide transfer (with many recipients), what happens on restart?"

I think they would attempt to be transferred, sequentially, and any issues with proof upload at that point would block the caretaker startup. Maybe I'm wrong about which errors would cause which functions to block though.

Reply:
We already spin up a goroutine for each individual parcel: https://github.com/lightninglabs/taproot-assets/blob/9ac1baacd996c20ce6b9c51fa77b4b84e0d3fe5d/tapfreighter/chain_porter.go#L337
So at startup, because the main event loop already is a goroutine and just starts another one for each parcel, we can feed in the parcels to resume synchronously and block on that.

Reply:
Ah, so we can handle parcels in parallel then?

In that case, IIUC, the only startup delay would be calling `advanceState()` for each resumed parcel, which should be fast.

And then we actually don't need to modify the current behavior, and the problem I was thinking of doesn't exist.

I'm not sure if we have an existing test that handles multiple parcels at once, that would be good to have to validate this.